### DISCUSSION TITLE
Undefined constant AMQP_NOPARAM in AmqpStamp

### DISCUSSION THREAD

**Discussion by musahcoding** [symfony/symfony#54761] [Q&A] [answered]
### Symfony version(s) affected

6.3.12

### Description

I just tried to use this package `symfony/amqp-messenger@6.3.12` last night and today noticed that I am receiving this exception:
```
"message": "Uncaught PHP Exception Error: \"Undefined constant \"AMQP_NOPARAM\"\" at /var/www/html/modeer-web-app/vendor/symfony/amqp-messenger/Transport/AmqpStamp.php line 27",
  "context": {
    "exception": {
      "class": "Error",
      "message": "Undefined constant \"AMQP_NOPARAM\"",
      "code": 0,
      "file": "/var/www/html/modeer-web-app/vendor/symfony/amqp-messenger/Transport/AmqpStamp.php:27"
    }
  },
  "level": 500,
  "level_name": "CRITICAL",
  "channel": "request",
  "datetime": "2024-04-20T17:02:33.166481+04:30",
  "extra": {}
}
```

And when I search for the definition of the constant `AMQP_NOPARAM` in `vendor/symfony/amqp-messenger`, it seems it is not defined. To do a workaround I created this file at `scripts/amqp-constants.php`:
```
<?php

if (!defined('AMQP_NOPARAM')) {
    define('AMQP_NOPARAM', 0);
}
```

and added the following to `composer.json`:
```
    "autoload": {
        "files": ["scripts/amqp-constants.php"],
```
But it does not really look the right thing for me, I must have been missing something or is it a bug?

### How to reproduce

I mentioned all the code above.

### Possible Solution

I added my workaround above, maybe it could also be moved to the repo itself?

**Comment by derrabus** [comment]
> 6.3.12

Note that this Symfony version [is not maintained anymore](https://symfony.com/releases/6.3). Even if you've identified a bug, you won't get any bugfixes unless you upgrade to a newer version.

> I must have been missing something or is it a bug?

[The AMQP extension](https://pecl.php.net/package/amqp) might be the something that you're missing.

**Comment by musahcoding** [comment]
Hey 👋
Thanks for the upgrade tip, I can do it if the problem will be fixed with
it, but since it is a live and complex project it will take some time.

About AMQP extension I already have. Indeed with my quick fix push and
consumption work normally.

Thanks a lot.

**Comment by derrabus** [comment]
But the constant that you're missing should be defined by the AMQP extension. Maybe you're using an old version of it?

**Comment by musahcoding** [comment]
Ok, I'll check it out.

**Comment by musahcoding** [comment]
@derrabus I am sorry for my above and incorrect reply. I do not have the pecl extension but on Ubuntu 22.04 I have `librabbitmq-dev` and `php-amqp` which I insatlled via `apt`. From the sound of it, these packages do not have that constant :thinking:

**Comment by discordier** [comment]
@musahcoding having them installed is not enough.
You need to also enable the extension via php.ini (usually something like `php-enmod` or something along these lines in Ubuntu).

You can check via `phpinfo()` if the `amqp` extension is loaded.

**Comment by musahcoding** [comment]
Thanks for the continued support, I think it is loaded, as I mentioned with that hack publish and consumption work, but also the following commands can confirm it too that the extension is loaded:
```
modeer_devops@lamp-ubuntu-s-2vcpu-2gb-nyc3-01:~$ php -i | grep rabbitmq
librabbitmq version => 0.10.0
modeer_devops@lamp-ubuntu-s-2vcpu-2gb-nyc3-01:~$ php -i | grep amqp
/etc/php/8.2/cli/conf.d/20-amqp.ini,
amqp
amqp.auto_ack => 0 => 0
amqp.cacert => no value => no value
amqp.cert => no value => no value
amqp.channel_max => 256 => 256
amqp.connect_timeout => 0 => 0
amqp.deserialization_depth => 128 => 128
amqp.frame_max => 131072 => 131072
amqp.global_prefetch_count => 0 => 0
amqp.global_prefetch_size => 0 => 0
amqp.heartbeat => 0 => 0
amqp.host => localhost => localhost
amqp.key => no value => no value
amqp.login => guest => guest
amqp.password => guest => guest
amqp.port => 5672 => 5672
amqp.prefetch_count => 3 => 3
amqp.prefetch_size => 0 => 0
amqp.read_timeout => 0 => 0
amqp.rpc_timeout => 0 => 0
amqp.sasl_method => AMQP_SASL_METHOD_PLAIN => AMQP_SASL_METHOD_PLAIN
amqp.serialization_depth => 128 => 128
amqp.timeout => no value => no value
amqp.verify => 1 => 1
amqp.vhost => / => /
amqp.write_timeout => 0 => 0
```

Somehow it seems that the apt implementation is different that pecl one. Even if that is the case, I would be happy to follow the pecl installation, but maybe you/we can emphasize in the symfony doc to use the pecl installation or else developers would run into this issue? :thinking:

**Comment by xabbuh** [comment]
Can you also share the output of the following commands?

```
php -r "var_dump(AMQP_NOPARAM);"
php -r "echo phpversion('amqp');"
```

**Comment by musahcoding** [comment]
Hey!
Here they are:
```
modeer_devops@lamp-ubuntu-s-2vcpu-2gb-nyc3-01:~$ php -r "var_dump(AMQP_NOPARAM);"
int(0)
modeer_devops@lamp-ubuntu-s-2vcpu-2gb-nyc3-01:~$ php -r "echo phpversion('amqp');"
2.1.2
```

**Comment by derrabus** [comment]
Ubuntu ships the old v1 of the extension. I suspect that v1 does not have the constant (which is just an alias for zero anyway). So the fix might be that we're using `0` instead of the constant in Symfony.

**Comment by xabbuh** [comment]
So this proves that the constant is defined. Could it be that you run PHP in different contexts (CLI and web for example) and that the other context (the request channel  in the log excerpt leads me to this assumption) doesn't have the extension enabled?

**Comment by derrabus** [comment]
> on Ubuntu 22.04 I have `librabbitmq-dev` and `php-amqp` which I insatlled via `apt`.

> ```
> modeer_devops@lamp-ubuntu-s-2vcpu-2gb-nyc3-01:~$ php -r "var_dump(AMQP_NOPARAM);"
> int(0)
> modeer_devops@lamp-ubuntu-s-2vcpu-2gb-nyc3-01:~$ php -r "echo phpversion('amqp');"
> 2.1.2
> ```

This does not add up: Ubunto ships v1: https://packages.ubuntu.com/jammy/php-amqp

**Comment by musahcoding** [comment]
@xabbuh I have already checked both web and CLI they use the same extensions.

@derrabus I am using using LAMP image of digital oceans where they have references to more recent versions of PHP 8.3, I see `ondrej` is also in the list of my sources which has the updates of PHP.

 ```
root@lamp-ubuntu-s-2vcpu-2gb-nyc3-01:/home/modeer_devops# ls /etc/apt/sources.list.d/ -l
total 32
-rw-r--r-- 1 root root  130 Oct 18  2023 digitalocean-agent.list
-rw-r--r-- 1 root root  129 Oct 18  2023 droplet-agent.list
-rw-r--r-- 1 root root   58 Apr 20 00:48 focal-security.list
-rw-r--r-- 1 root root  100 Oct 18  2023 nodesource.list
-rw-r--r-- 1 root root  140 Sep  7  2023 ondrej-ubuntu-php-jammy.list
-rw-r--r-- 1 root root 1285 Apr 20 01:02 rabbitmq.list
-rw-r--r-- 1 root root  417 Apr 20 00:49 rabbitmq_rabbitmq-server.list
-rw-r--r-- 1 root root  428 Oct 18  2023 symfony-stable.list
```

**Comment by xabbuh** [comment]
```php
<?php

var_dump(AMQP_NOPARAM);
var_dump(phpversion('amqp'));
```

What is the output of this script when you access it over your web server?

**Comment by musahcoding** [comment]
Hey @xabbuh :wave: here it is:
```
modeer_devops@lamp-ubuntu-s-2vcpu-2gb-nyc3-01:~$ php test.php 
int(0)
string(5) "2.1.2"
modeer_devops@lamp-ubuntu-s-2vcpu-2gb-nyc3-01:~$ cat test.php 
<?php

var_dump(AMQP_NOPARAM);
var_dump(phpversion('amqp'));
```

**Comment by discordier** [comment]
I think he meant you should access it via web SAPI (via browser, curl, ...) instead of CLI SAPI.

**Comment by xabbuh** [comment]
@discordier is right

**Comment by derrabus** [comment]
I think, I'm moving this to support now to reduce the noise on the bug tracker.

**Comment by musahcoding** [comment]
![image](https://github.com/symfony/symfony/assets/77657445/8da58de7-829d-421b-b8bc-cad8e4f297cc)
I am sorry :smile: here is the output from browser

**Reply by musahcoding** [reply]
@discordier Having a life of contributors I know you are very busy. Just to make sure you have not missed the previous notification, thanks 👍🏼

**Reply by discordier** [reply]
I missed it.

Your sceenshot proves that the constant is defined.

So either:
- the problem does not exist anymore
- you are using a different SAPI than on your original message above (you should use the exact same webspace)

**Reply by musahcoding** [reply]
@discordier Thanks for all the way down here with the investigation. I merged this MR:
![image](https://github.com/symfony/symfony/assets/77657445/8a682fcb-c0d2-4172-9731-497d746386a1)

Since this script should be run on every request, I should get this exception if the constant was not defined. So you are right, something has changed now and the original question is not valid anymore.

**Comment by musahcoding** [comment, marked as answer]
Hi :wave: I found the issue finally :tada: 

**TLDR:**
When I upgraded the PHP version of server from 8.2 to 8.3, I did not purge the `php8.2*` packages by running `sudo apt purge php8.2*`. Doing so fixed my problem of `undefined AMQP_NOPARAM`

<hr>

**Details:**

I was working with one of the old images of the server, and I noticed that it runs on an out dated version of PHP 8.2, where the server runs on PHP 8.3, so I did:

```
sudo apt update
sudo apt upgrade -y
```
And now I had:
```
$ php -v
PHP 8.3.9 (cli) (built: Jul  5 2024 12:03:46) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.3.9, Copyright (c) Zend Technologies
    with Zend OPcache v8.3.9, Copyright (c), by Zend Technologies
```

But then the deployment failed again with the same error `AMQP_NOPARAM not found` :thinking: 

Then after googling again + gemini, I suspected the version conflict, and running `sudo apt purge php8.2*` removed the old PHP 8.2 modules including the `php-amqp` extension. 

Now deployment worked again :tada:

**Comment by discordier** [comment]
Am I the only one failing to see any relevance of the memcached extension when this was all about AMQP?

**Reply by musahcoding** [reply]
Nope my mistake 🫠 It was php-amqp. Fixed the other comment too
